controllers/postgresqluser_controller.go

```diff
@@ -98,26 +115,69 @@ func (r *PostgreSQLUserReconciler) reconcile(reqLogger logr.Logger, request reco
 		return ctrl.Result{}, err
 	}
 
-	// User instance created or updated
-	reqLogger = reqLogger.WithValues("user", user.Spec.Name, "rolePrefix", r.RolePrefix)
-	reqLogger.Info("Reconciling found PostgreSQLUser resource", "user", user.Spec.Name)
-
-	// Error check in the bottom because we want aws policy to be set no matter what.
-	granterErr := r.Granter.SyncUser(reqLogger, request.Namespace, r.RolePrefix, *user)
-
 	var awsCredentials *credentials.Credentials
 	if len(r.AWSProfile) != 0 {
 		awsCredentials = credentials.NewSharedCredentials("", r.AWSProfile)
 	} else {
 		awsCredentials = credentials.NewStaticCredentials(r.AWSAccessKeyID, r.AWSSecretAccessKey, "")
 	}
 
-	awsPolicyErr := r.SetAWSPolicy(reqLogger, awsCredentials, iam.AddUserConfig{
+	awsConfig := &aws.Config{
+		Region:      aws.String(r.AWSRegion),
+		Credentials: awsCredentials,
+	}
+
+	// Initialize session to AWS
+	session, err := session.NewSession(awsConfig)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("session initialization for region %s: %w", r.AWSRegion, err)
+	}
+
+	client := iam.NewClient(session, reqLogger, r.AWSAccountID, r.IAMPolicyPrefix)
+
+	markedToBeDeleted := user.GetDeletionTimestamp() != nil
+	if markedToBeDeleted {
+		if inList(user.Finalizers, userFinalizer) {
+			// Run finalization logic for userFinalizer. If the
+			// finalization logic fails, don't remove the finalizer so
+			// that we can retry during the next reconciliation.
+			if err := r.finalizeUser(reqLogger, client, user); err != nil {
+				return ctrl.Result{}, err
+			}
+
+			// Remove finalizer. Once all finalizers have been
+			// removed, the object will be deleted.
+			controllerutil.RemoveFinalizer(user, userFinalizer)
+			err := r.Update(context.TODO(), user)
+			if err != nil {
+				return ctrl.Result{}, err
+			}
+		}
+		return ctrl.Result{}, nil
```

Comment:
Small thing, but can we invert the check to avoid the nesting?

```suggestion
		if !inList(user.Finalizers, userFinalizer) {
			return ctrl.Result{}, nil
		}
		// Run finalization logic for userFinalizer. If the
		// finalization logic fails, don't remove the finalizer so
		// that we can retry during the next reconciliation.
		if err := r.finalizeUser(reqLogger, client, user); err != nil {
			return ctrl.Result{}, err
		}

		// Remove finalizer. Once all finalizers have been
		// removed, the object will be deleted.
		controllerutil.RemoveFinalizer(user, userFinalizer)
		err := r.Update(context.TODO(), user)
		if err != nil {
			return ctrl.Result{}, err
		}

		return ctrl.Result{}, nil
```